Add support for Elasticsearch 2.x

ael-code opened the pull request:
You can find the old discussion about ES `2.x` support on old PR #267.

This PR aims to make the code completely compatible with Elasticsearch 2.x versions. 

A lot of functionalities have been deprecated and even removed in the new versions of ES, thus I needed to change a lot of stuff.

Even if I made a lot of small changes I've leaved the tests almost unchanged, moreover I've made travis run our tests also upon last version of Elasticsearch 2.x.

ael-code commented:
This PR do not depends any more on #280

boyska commented:
I've had a look at the code. It seems fine, but I am really missing how does it handle migration: there is a warning when the mapping changes, but what should the user do about it? will everything still work?
(yes, I know, I promised to work about testing migrations, but I didn't)

ael-code commented:
On 04/08/2016 04:37 PM, BoySka wrote:
> I've had a look at the code. It seems fine, but I am really missing how
> does it handle migration: there is a warning when the mapping changes,
> but what should the user do about it? will everything still work?
> (yes, I know, I promised to work about testing migrations, but I didn't)
> 

It was a little bit hard but finally I've managed to expose a new functionality: `librenat-db reindex`.

The internal function is DB.reindex() and this is its docstring:

```
Rebuilt the current index

This function could be useful in the case you want to change some index settings/mappings and you don't want to loose all the entries belonging to that index.

This function is built in such a way that you can continue to use the old index name, this is achieved using index aliases.

The old index will be cloned into a new one with the given `index_conf`.
If we are working on an alias, it is redirected to the new index.
Otherwise a brand new alias with the old index name is created in such a way that points to the newly create index.

Keep in mind that even if you can continue to use the same index name, the old index will be deleted.
```

Thanks to this new tool we can start to think about a migration procedure and how to test it.

ael-code commented:
> Thanks to this new tool we can start to think about a migration procedure and how to test it.
Ok, I've tested the following upgrade procedure.
 - Install libreant v0.4 and elasticsearch 1.x.
 - Create a database and add some volumes.

      (I've tested one with 6766 volumes)
 - Upgrade to the new elasticsearch/libreant version.
      (I've used elasticsearch v2.3.1 and the code of this PR for libreant)
      Don't forget to upgrade also py-elasticsearch to the new version with:
      ```
      sed -i "s/\('elasticsearch\)\(.*\)\('\)/\1 >2\3/g" setup.py
      pip install --upgrade elasticsearch>=2.0
      ```
     
 - Run `libreant-db reindex`
 - Launch libreant

Seems a good test case to you? Can you provide/tests some others real cases?

ael-code commented:
I'm going to merge this tomorrow evening. If you have some objections please let me know.

boyska commented:
The test case is good, but what happens if the sysadmin upgrades both elasticsearch and elasticsearch-py but does not run the `reindex`? I expect that nothing should bring the db to a "mixed" state: either operations are fine, or they fail explicitly.

ael-code commented:
On 04/26/2016 03:56 PM, BoySka wrote:
> The test case is good, but what happens if the sysadmin upgrades both
> elasticsearch and elasticsearch-py but does not run the |reindex|? I
> expect that nothing should bring the db to a "mixed" state: either
> operations are fine, or they fail explicitly.

You can make the downgrade without re-indexing, but you can't  make the
upgrade without performing the reindex action.

boyska commented:
ael:
>On 04/26/2016 03:56 PM, BoySka wrote:
>> The test case is good, but what happens if the sysadmin upgrades both
>> elasticsearch and elasticsearch-py but does not run the |reindex|? I
>> expect that nothing should bring the db to a "mixed" state: either
>> operations are fine, or they fail explicitly.
>
>You can make the downgrade without re-indexing, but you can't  make the
>upgrade without performing the reindex action.

downgrade? My usecase is about a sysadmin not doing anything.

ael-code commented:
On 04/27/2016 02:36 PM, BoySka wrote:
> ael:
>>On 04/26/2016 03:56 PM, BoySka wrote:
>>> The test case is good, but what happens if the sysadmin upgrades both
>>> elasticsearch and elasticsearch-py but does not run the |reindex|? I
>>> expect that nothing should bring the db to a "mixed" state: either
>>> operations are fine, or they fail explicitly.
>>
>>You can make the downgrade without re-indexing, but you can't make the
>>upgrade without performing the reindex action.
> 
> downgrade? My usecase is about a sysadmin not doing anything.
> 

If you upgrade and you did not perform the reindex, libreant will
interact with elasticsearch and will trigger some ES exception caused by
either some deprecated index settings or some deprecated functions.

boyska commented:
ael:
>On 04/27/2016 02:36 PM, BoySka wrote:
>> ael:
>>>On 04/26/2016 03:56 PM, BoySka wrote:
>>>> The test case is good, but what happens if the sysadmin upgrades both
>>>> elasticsearch and elasticsearch-py but does not run the |reindex|? I
>>>> expect that nothing should bring the db to a "mixed" state: either
>>>> operations are fine, or they fail explicitly.
>>>
>>>You can make the downgrade without re-indexing, but you can't make the
>>>upgrade without performing the reindex action.
>>
>> downgrade? My usecase is about a sysadmin not doing anything.
>
>If you upgrade and you did not perform the reindex, libreant will
>interact with elasticsearch and will trigger some ES exception caused by
>either some deprecated index settings or some deprecated functions.

ok, failing hard is good. I mean, is better than failing silently and
creating an inconsistent DB.

Could we handle those exceptions showing a clear description of what to
do next, like "thaks for upgrading, but now your db needs a reindex,
too" ?

boyska commented on travis/install_dependencies.sh:
```diff
@@ -0,0 +1,11 @@
+    echo "deb http://packages.elastic.co/elasticsearch/2.x/debian stable main" | sudo tee -a /etc/apt/sources.list.d/elk.list
+    sudo apt-get update && sudo apt-get install elasticsearch -y
+    sudo service elasticsearch start
+    sed -i "s/\('elasticsearch\)\(.*\)\('\)/\1 >2\3/g" setup.py
```
what a `sed` expression! what does it do?

ael-code replied:
On 04/27/2016 08:27 PM, BoySka wrote:
> In travis/install_dependencies.sh
> <https://github.com/insomnia-lab/libreant/pull/281#discussion_r61311207>:
> 
>> @@ -0,0 +1,11 @@
>> +#! /bin/bash
>> +
>> +if [[ $VERSION_ES == "2.x" ]]; then
>> +    sudo apt-get autoremove --purge elasticsearch
>> +    wget -qO - https://packages.elastic.co/GPG-KEY-elasticsearch | sudo apt-key add -
>> +    echo "deb http://packages.elastic.co/elasticsearch/2.x/debian stable main" | sudo tee -a /etc/apt/sources.list.d/elk.list
>> +    sudo apt-get update && sudo apt-get install elasticsearch -y
>> +    sudo service elasticsearch start
>> +    sed -i "s/\('elasticsearch\)\(.*\)\('\)/\1 >2\3/g" setup.py
> 
> what a |sed| expression! what does it do?
> 

It simply change the ealsticsearch dependency version in `setup.py` from
`elasticsearch >=1, <2` to `elasticsearch >2`.

ael-code commented:
On 04/27/2016 06:48 PM, BoySka wrote:
> ael:
>>On 04/27/2016 02:36 PM, BoySka wrote:
>>> ael:
>>>>On 04/26/2016 03:56 PM, BoySka wrote:
>>>>> The test case is good, but what happens if the sysadmin upgrades both
>>>>> elasticsearch and elasticsearch-py but does not run the |reindex|? I
>>>>> expect that nothing should bring the db to a "mixed" state: either
>>>>> operations are fine, or they fail explicitly.
>>>>
>>>>You can make the downgrade without re-indexing, but you can't make the
>>>>upgrade without performing the reindex action.
>>>
>>> downgrade? My usecase is about a sysadmin not doing anything.
>>
>>If you upgrade and you did not perform the reindex, libreant will
>>interact with elasticsearch and will trigger some ES exception caused by
>>either some deprecated index settings or some deprecated functions.
> 
> ok, failing hard is good. I mean, is better than failing silently and
> creating an inconsistent DB.
> 
> Could we handle those exceptions showing a clear description of what to
> do next, like "thaks for upgrading, but now your db needs a reindex,
> too" ?

It's difficult to check at runtime if libreant have been just upgraded.

At the moment when the first connection to the database is issued, we
check if installed mappings for the current index are different from the
one provided by the current version of libreant. If this is the case we
output the following warning:

```
An old or wrong mapping has been found for index: "{0}"
```

Do you prefer another message? Do you want to proceed in another way?

boyska commented:
ael:
>On 04/27/2016 06:48 PM, BoySka wrote:
>> ok, failing hard is good. I mean, is better than failing silently and
>> creating an inconsistent DB.
>>
>> Could we handle those exceptions showing a clear description of what to
>> do next, like "thaks for upgrading, but now your db needs a reindex,
>> too" ?
>
>It's difficult to check at runtime if libreant have been just upgraded.
>
>At the moment when the first connection to the database is issued, we
>check if installed mappings for the current index are different from the
>one provided by the current version of libreant. If this is the case we
>output the following warning:
>
>```
>An old or wrong mapping has been found for index: "{0}"
>```

the fact that it is logged is good, but I wonder whether we should
handle it at the UI level, too. I mean, what will the user see after
this? Will the query work? If exceptions are raised, I expect them to be
catched and informative.

Something like "this libreant installation is upgrading; please complete
your upgrade process. Visit
https://libreant.readthedocs.io/en/latest/sysadmin.html#upgrading to
know more"

I am sorry but a good testing of this PR is hard to do, so I'll need
some time to do it. If someone else is faster than me, that's great.

>Do you prefer another message? Do you want to proceed in another way?

A more informative message would be great, explaining that if the
sysadmin just upgraded, this is expected, nothing to worry about, but
`reindex` must be run. Maybe linking to documentation would be a good
way of doing it.

ael-code commented:
I've changed the approach to the upgrade procedure. Please hold on for the PR update

ael-code commented:
### Index mappings update

Every time libreant starts an attempt to update the index mappings is done, if it fails, an error with detailed description is logged:
```
2016-05-01 12:39:00,365 [libreantdb.api] [ERROR] Couldn't update index properties mapping: { '_language': { Mapper for [_language] conflicts with existing mapping in other types: [mapper [_language] has different [index] values}}
```

Followed by a warning log:
```
2016-05-01 12:39:00,365 [libreantdb.api] [WARNING] An old or wrong properties mapping has been found for index: "propirata", this could led to some errors. It is recomanded to perform a reindexing
```

### Migration of timestamp field

Since Elasticsearch v2.0 the `_timestamp` field has been deprecated. At the moment we use this field only for the 'recently added page'. Thanks to the code in the first commit I've replaced the bultin special `_timestamp` field with a normal `_insertion_date` field.

Unfortunately the old `_timestamp` field was a special field, thus it is not possible to use the reindex procedure to perform the migration from the old field to the new one.

I've written a specific tool to perform this migration:
https://github.com/ael-code/libreant/blob/timestamp_migration/libreant_migrate_timestamp.py
It's a simple command line tool that given an index it copies the value of the special `_timestamp` field into the new `_insertion_date` for each volumes in that index.


### Upgrade flow
I know... this is a mess. Let's make some order.

In the following scenarios it has been assumed that you have an already populated
database and you want to preserve it.

#### Upgrade only libreant

 - run the `libreant_migrate_timestamp` tool
 - upgrade libreant
 - Perform a reindex.

##### Notes
If you don't perform the first step, libreant will start but the `/recents` page will trigger a RequestError with the follwing reason:
```
Parse Failure [No mapping found for [_insertion_date] in order to sort on]
```

If you don't perform the last step (reindexing) libreant will continue to work but you could in some really rare cases have some strange results in the `related volumes` section.

#### Upgrade both elasticsearch and libreant

 - run the `libreant_migrate_timestamp` tool
 - upgrade libreant/elasticsearch
 - Perform a reindex.

##### Notes
If you don't perform the first step you will loose the insertion date information about all the already indexed volumes. The good thing is that libreant will continue to work, the only problem is that the 'recently added page' will show the label `invalid date` instead of the insertion date of the volumes.

If you don't perform the last step (reindexing) libreant will continue to work but you could in some really rare cases have some strange results in the `related volumes` section.

boyska commented:
mh, can I run the migration thing AFTER having upgraded? It is the only feasible workflow for me. Let's say that I am using the last release of libreant. I can't run the migration tool, because I don't have it. I will first of all upgrade the code, then I can run it.

Second question: is the migration tool idempotent? That is, what if I run it mutiple times? Knowing that it can be run multiple times would make its usage less problematic.

Third question: if I upgrade libreant, can I still use elasticsearch 1, right?

ael-code commented:
On 05/01/2016 06:38 PM, BoySka wrote:
> mh, can I run the migration thing AFTER having upgraded?It is the only
> feasible workflow for me.

You can launch the migration tool as long as you remain with a version
of elasticsearch < 2.0.

> Let's say that I am using the last release of
> libreant. I can't run the migration tool, because I don't have it. I
> will first of all upgrade the code, then I can run it.
> 

The migration tool is a standalone script, independent from the libreant
code, and we need to make possible to easily donwload it.

With future changes of the mappings we will use the `libreant-db
reindex` tool, but unfortunately for this special case we need some
special procedure.


> Second question: is the migration tool idempotent? That is, what if I
> run it mutiple times? Knowing that it can be run multiple times would
> make its usage less problematic.
> 

Yes, it's idempotent. You can run it multiple time.

> Third question: if I upgrade libreant, can I still use elasticsearch 1,
> right?


I've explained in details this scenario under *Upgrade only libreant*.
If it's not clear I can try to explain it in a better way.


Do we want to open a repo for the migration tool?

boyska commented:
>On 05/01/2016 06:38 PM, BoySka wrote:
>> mh, can I run the migration thing AFTER having upgraded?It is the only
>> feasible workflow for me.
>
>You can launch the migration tool as long as you remain with a version
>of elasticsearch < 2.0.

Ok, this is great

>> Let's say that I am using the last release of
>> libreant. I can't run the migration tool, because I don't have it. I
>> will first of all upgrade the code, then I can run it.
>>
>
>The migration tool is a standalone script, independent from the
>libreant code, and we need to make possible to easily donwload it.

mh, don't like this approach: putting it inside libreant makes much more
sense. Maybe not inside `libreant-db` but in a separate script? that
would be reasonable.

I'd expect that an upgrade also gives the tools needed to handle it
correctly. So I think that we should bundle the tool in libreant.
Otherwise the sysadmin will discover, after having done the upgrade,
that the upgrade is half-broken, and the way to fix requires downloading
and installing one more thing.

>With future changes of the mappings we will use the `libreant-db
>reindex` tool, but unfortunately for this special case we need some
>special procedure.

Yes this is fine

>> Third question: if I upgrade libreant, can I still use elasticsearch 1,
>> right?
>
>I've explained in details this scenario under *Upgrade only libreant*.
>If it's not clear I can try to explain it in a better way.

It was clear, but I was asking for a confirmation :)
this+idempotence make the upgrade path definitely easier

>Do we want to open a repo for the migration tool?

Nah, as explained above, please keep everything simple to
install/package. You might be thinking: "a migration tool for the
previous version is not part of the core set of things needed for
libreant to run" but while this is theoretically correct, I find it a
bit exaggerated for such a simple script: it would require much more
documentation to be written, integrated, etc.

ael-code commented:
I've simplified a lot the update procedure:

I've dropped the `libreant-db reindex` command that was only used to update the index mappings.
Now we have the brand new `libreant-db upgrade`.
This command will take care of the entire process of migrating/updating the database in order to keep it aligned with the new versions of libreant.

For the moment it performs the follwing step:
 - recognizes the necessity to perform a migration of the `_timestamp` field
   and ask permission to the user to perform it.

 - recognizes if the index mappings are old/wrong and ask permission to
   the user to perform a complete reindex in order to update them.

## Upgrade flow
The ugrade flow now it's a lot easier and it's the same in the case you upgrade only libreant or both libreant and elasticsearch. 
In the following scenarios it has been assumed that you have an already populated
database and you want to preserve it.

#### Upgrade only libreant

 - upgrade libreant
 - run `libreant-db upgrade`

##### Notes
If you don't perform the upgrade of the database, libreant will start but the `/recents` page will trigger a RequestError and in some really rare cases you could have some strange results in the `related volumes` section.

#### Upgrade both elasticsearch and libreant

 - upgrade libreant/elasticsearch
 - run `libreant-db upgrade`

##### Notes
If you don't perform the upgrade of the database you will loose the insertion date information about all the already indexed volumes. The good thing is that libreant will continue to work, the only problem is that the 'recently added page' will show the label `invalid date` instead of the insertion date of the volumes.
Moreover if you don't perform the upgrade of the database in some really rare cases you could have some strange results in the `related volumes` section.

## One more good news
In a previous comment I've said
> You can launch the migration tool as long as you remain with a version of elasticsearch < 2.0.

but I've modified a little bit the code in order to make possible to launch the new `libreant-db upgrade` tool also after that elasticsearch has been upgraded to a version > 2.0 and even restarted.

boyska commented:
ael:
>I've simplified a lot the update procedure:

Your description seems great, just what I would have liked it to be.
I need to test it, though.
I'll promise that I'll *start* doing so (and maybe giving feedback) in
the next hours. If you have any, even simple, script, trick or other
suggestion, please share me (even privately is ok).

>Now we have the brand new `libreant-db upgrade`.
>This command will take care of the entire process of migrating/updating the database in order to keep it aligned with the new versions of libreant.
>
>For the moment it performs the follwing step:
> - recognizes the necessity to perform a migration of the `_timestamp` field
>   and ask permission to the user to perform it.
>
> - recognizes if the index mappings are old/wrong and ask permission to
>   the user to perform a complete reindex in order to update them.

very nice. Can we have something (either a separate but related
subcommand, or an option) that can tell a sysadmin if an upgrade is
needed, without actually doing it, nor being prompted? This would be
nice for scripts like

`libreant-db update-needed && mail root -s "update" <<< really,update`

or to test it

```sh
libreant insert-volume -l it - <<<'{}'
git pull && python setup.py install
libreant-db update-needed || echo "It should have been needed"
libreant-db update
libreant-db update-needed && echo "It should not be needed anymore"
```

This can be postponed of course.

>In a previous comment I've said
>> You can launch the migration tool as long as you remain with a
>> version of elasticsearch < 2.0.
>
>but I've modified a little bit the code in order to make possible to
>launch the new `libreant-db upgrade` tool also after that elasticsearch
>has been upgraded to a version > 2.0 and even restarted.

what do you mean with "even restarted"?

boyska commented:
Random notes while I keep testing:
* after upgrading only libreant, the `RequestError` is reported only sometimes (rarely, actually), even though the message complaining about mappings is always present in the log.
* `libreant-db upgrade` asks me twice if I really want to upgrade. Yes, I run upgrade because I want to upgrade :) I think that it should at least provide an option not to bother me. I also don't understand why the first question defaults to yes, while the second defaults to no.

ael-code commented:
On 05/10/2016 12:12 AM, BoySka wrote:
> ael:
>>I've simplified a lot the update procedure:
> 
> Your description seems great, just what I would have liked it to be.
> I need to test it, though.
> I'll promise that I'll *start* doing so (and maybe giving feedback) in
> the next hours. If you have any, even simple, script, trick or other
> suggestion, please share me (even privately is ok).
> 
>>Now we have the brand new `libreant-db upgrade`.
>>This command will take care of the entire process of migrating/updating
> the database in order to keep it aligned with the new versions of libreant.
>>
>>For the moment it performs the follwing step:
>> - recognizes the necessity to perform a migration of the `_timestamp`
> field
>> and ask permission to the user to perform it.
>>
>> - recognizes if the index mappings are old/wrong and ask permission to
>> the user to perform a complete reindex in order to update them.
> 
> very nice. Can we have something (either a separate but related
> subcommand, or an option) that can tell a sysadmin if an upgrade is
> needed, without actually doing it, nor being prompted? This would be
> nice for scripts like
> 
> `libreant-db update-needed && mail root -s "update" <<< really,update`
> 
> or to test it
> 
> ```sh
> libreant insert-volume -l it - <<<'{}'
> git pull && python setup.py install
> libreant-db update-needed || echo "It should have been needed"
> libreant-db update
> libreant-db update-needed && echo "It should not be needed anymore"
> ```
> 
> This can be postponed of course.
> 

I've added the -c/--check-only flag. With the following help message:
```
Does not perform any action on database, it will exit with code 123 if
an upgraded is needed, and 0 if it's not. Any other exit code means that
an error occurred.

>>In a previous comment I've said
>>> You can launch the migration tool as long as you remain with a
>>> version of elasticsearch < 2.0.
>>
>>but I've modified a little bit the code in order to make possible to
>>launch the new `libreant-db upgrade` tool also after that elasticsearch
>>has been upgraded to a version > 2.0 and even restarted.
> 
> what do you mean with "even restarted"?
> 

At least on debian, when you upgrade the elasticsearch package it does
not change the runtime version till you restart the daemon.

So with 'even restarted' I simply mean that the upgrade tool works with
a running version of elasticsearch >= 2.0.

ael-code commented:
On 05/10/2016 05:41 AM, BoySka wrote:
> Random notes while I keep testing:
> 
>   * after upgrading only libreant, the |RequestError| is reported only
>     sometimes (rarely, actually), even though the message complaining
>     about mappings is always present in the log.

It does not seems the same to me. Can you provide the case in which the
RequestError is not raised.

>   * |libreant-db upgrade| asks me twice if I really want to upgrade.
>     Yes, I run upgrade because I want to upgrade :) 

I've built the command line utility in such a way that help you to fix
some db problems. From my point of view the '_timestamp' and the one
about mappings are two different and separate issues.
It could be the case (development purposes) that you want to fix the
first issue but to keep the old mappings.

>     I think that it should at least provide an option not to bother me.

what about `yes | libreant upgrade` ?

>     I also don't understand why the first question defaults to yes, while the second
>     defaults to no.

I've update the default answers both to False.

boyska commented:
ael:
>On 05/10/2016 05:41 AM, BoySka wrote:
>> Random notes while I keep testing:
>>
>>   * after upgrading only libreant, the |RequestError| is reported only
>>     sometimes (rarely, actually), even though the message complaining
>>     about mappings is always present in the log.
>
>It does not seems the same to me. Can you provide the case in which the
>RequestError is not raised.

I just tested with `/recents`. It's just random. At first `RequestError`
was happening, but then it was not. I could try to test this better.

>>   * |libreant-db upgrade| asks me twice if I really want to upgrade.
>>     Yes, I run upgrade because I want to upgrade :)
>
>I've built the command line utility in such a way that help you to fix
>some db problems. From my point of view the '_timestamp' and the one
>about mappings are two different and separate issues.
>It could be the case (development purposes) that you want to fix the
>first issue but to keep the old mappings.
>
>>     I think that it should at least provide an option not to bother me.
>
>what about `yes | libreant upgrade` ?

that's what I do, I was just complaining that this is horrible, because
it doesn't distinguish between what is expected and what is unexpected.
Don't give much importance to this rant on usability and user input, but
when I see that a program is asking for my input, I think that:
* there's a choice that a machine really can't do on its own because
  it's up to me (like choosing the hostname during installation)
* I'm about to do a dangerous thing
* the application wants to notify me of something not exactly expected

I think that none of those applies here. My favourite style for this
kind of stuff is [how apt-get does -y](https://manned.org/apt-get.8):

```
-y, --yes, --assume-yes
   Automatic yes to prompts; assume "yes" as answer to all prompts and
   run non-interactively. If an undesirable situation, such as changing
   a held package, trying to install a unauthenticated package or
   removing an essential package occurs then apt-get will abort.
```

>>     I also don't understand why the first question defaults to yes,
>>     while the second defaults to no.
>
>I've update the default answers both to False.

So if you run `libreant-db upgrade` and follow the defaults, it won't be
upgraded. And while I can `yes | libreant-db upgrade` for testing sake,
I won't suggest any sysadmin to do so.

ael-code commented:
On 05/11/2016 10:18 PM, BoySka wrote:
> ael:
>>On 05/10/2016 05:41 AM, BoySka wrote:
>>> Random notes while I keep testing:
>>>
>>> * after upgrading only libreant, the |RequestError| is reported only
>>> sometimes (rarely, actually), even though the message complaining
>>> about mappings is always present in the log.
>>
>>It does not seems the same to me. Can you provide the case in which the
>>RequestError is not raised.
> 
> I just tested with `/recents`. It's just random. At first `RequestError`
> was happening, but then it was not. I could try to test this better.
> 
>>> * |libreant-db upgrade| asks me twice if I really want to upgrade.
>>> Yes, I run upgrade because I want to upgrade :)
>>
>>I've built the command line utility in such a way that help you to fix
>>some db problems. From my point of view the '_timestamp' and the one
>>about mappings are two different and separate issues.
>>It could be the case (development purposes) that you want to fix the
>>first issue but to keep the old mappings.
>>
>>> I think that it should at least provide an option not to bother me.
>>
>>what about `yes | libreant upgrade` ?
> 
> that's what I do, I was just complaining that this is horrible, because
> it doesn't distinguish between what is expected and what is unexpected.
> Don't give much importance to this rant on usability and user input, but
> when I see that a program is asking for my input, I think that:
> * there's a choice that a machine really can't do on its own because
> it's up to me (like choosing the hostname during installation)
> * I'm about to do a dangerous thing
> * the application wants to notify me of something not exactly expected
> 
> I think that none of those applies here. My favourite style for this
> kind of stuff is [how apt-get does -y](https://manned.org/apt-get.8):
> 
> ```
> -y, --yes, --assume-yes
> Automatic yes to prompts; assume "yes" as answer to all prompts and
> run non-interactively. If an undesirable situation, such as changing
> a held package, trying to install a unauthenticated package or
> removing an essential package occurs then apt-get will abort.
> ```

I think that this is the case of "I'm about to do a dangerous thing".
Anyway I've added the --yes flag as I already did in the #291.


Is there any other problem or is finally the case that we are going to
merge this?

boyska commented:
> I think that this is the case of "I'm about to do a dangerous thing".
> Anyway I've added the --yes flag as I already did in the #291.

great, will test it

> Is there any other problem or is finally the case that we are going to
> merge this?

I'd like to know if it's normal that I cant' always reproduce the RequestError.
At the moment I only tested the "upgrade only libreant" upgrade (and not even so well). I need to test the "upgrade everything together", still.

ael-code commented:
On 05/14/2016 12:04 AM, BoySka wrote:
>     I think that this is the case of "I'm about to do a dangerous thing".
>     Anyway I've added the --yes flag as I already did in the #291
>     <https://github.com/insomnia-lab/libreant/pull/291>.
> 
> great, will test it
> 
>     Is there any other problem or is finally the case that we are going to
>     merge this?
> 
> I'd like to know if it's normal that I cant' always reproduce the
> RequestError.

Is really difficult answer this question because you didn't provide any
information about neither the case in which the error is  raised nor the
case in which it won't.

> At the moment I only tested the "upgrade only libreant" upgrade (and not
> even so well). I need to test the "upgrade everything together", still.
>

I hope that it won't take too much time to perform the additional tests.
I hope to merge this soon.

boyska commented:
I'm testing the simultaneous upgrade of libreant and elasticsearch.
I see this problem: https://paste.debian.net/683955/ . That's what happens:
* install 1.7
* populate db
* upgrade everything
* `libreant-db upgrade --only-check` correctly says that it must be upgraded
* I do `curl localhost/recents` and...
** the elasticsearch log correctly complains about mapping problems
** still, the page gives 200 and it is rendered, except that the timestamp is empty

Is this correct? I would have expected some error to appear

boyska commented:
Same test suite of before, now I ignore the fact that doing `curl localhost/recents` doesn't report any error. I run `libreant-db upgrade -y`, which doesn't produce any output. I guess this means "everything is fine". However, right after it, `libreant-db upgrade --check-only` still reports exit code 123, meaning that the upgrade is still to do. What did I missed?

ael-code commented:
On 05/15/2016 02:57 AM, BoySka wrote:
> I'm testing the simultaneous upgrade of libreant and elasticsearch.
> I see this problem: https://paste.debian.net/683955/ . That's what happens:
> 
>   * install 1.7
>   * populate db
>   * upgrade everything
>   * |libreant-db upgrade --only-check| correctly says that it must be
>     upgraded
>   * I do |curl localhost/recents| and... ** the elasticsearch log
>     correctly complains about mapping problems ** still, the page gives
>     200 and it is rendered, except that the timestamp is empty
> 
> Is this correct? I would have expected some error to appear
> 

This is completely correct. In the case you cannot retrieve the
`_insertion_date` field you still can access the `/recents` page but
dates are replaced by the string `invalid_date`.

ael-code commented:
On 05/15/2016 03:00 AM, BoySka wrote:
> Same test suite of before, now I ignore the fact that doing |curl
> localhost/recents| doesn't report any error. I run |libreant-db upgrade
> -y|, which doesn't produce any output. I guess this means "everything is
> fine". However, right after it, |libreant-db upgrade --check-only| still
> reports exit code 123, meaning that the upgrade is still to do. What did
> I missed?
> 

This is weird. Can you provide additional details. For example you can
run all the above commands with the debug flag.

boyska commented:
Sure; added the debug flag, but there's not much more.
* [Logs](https://paste.debian.net/hidden/f4ecc9a9/)
* [test "suite"](https://gist.github.com/boyska/7984ac7982a6539d00eb5d37a0915af2)

boyska commented:
ael:
>boyska:
>> I'm testing the simultaneous upgrade of libreant and elasticsearch.
>> I see this problem: https://paste.debian.net/683955/ . That's what happens:
>>
>>   * install 1.7
>>   * populate db
>>   * upgrade everything
>>   * |libreant-db upgrade --only-check| correctly says that it must be
>>     upgraded
>>   * I do |curl localhost/recents| and... ** the elasticsearch log
>>     correctly complains about mapping problems ** still, the page gives
>>     200 and it is rendered, except that the timestamp is empty
>>
>> Is this correct? I would have expected some error to appear
>>
>
>This is completely correct. In the case you cannot retrieve the
>`_insertion_date` field you still can access the `/recents` page but
>dates are replaced by the string `invalid_date`.

as I said, there is no "invalid_date" string, the timestamp is empty. At
least in the HTML, I haven't checked if invalid_date is inserted through
JS (but that would be super weird).

ael-code commented:
> Sure; added the debug flag, but there's not much more.
> 
>    -  Logs
>    - test "suite"

I've tried hard, but I cannot understand which is the test case flow. Did it happen in `do-upgradeboth.sh`? Can you provide a detailed test flow?

boyska commented:
ael:
>> Sure; added the debug flag, but there's not much more.
>>
>>    - Logs
>>    - test "suite"
>
>I've tried hard, but I cannot understand which is the test case flow.
>Did it happen in `do-upgradeboth.sh`? Can you provide a detailed test
>flow?

sure. The numbered ones are thing I run on my system. The do-* ones are
run inside the container.

00base.sh, using do-base.sh, creates a base system for next
tests: installs the stable version, ES 1.7, and creates some documents.
the other tests are run on ephemeral containers, so they don't touch
each other. The typical sequence for testing is bash 00bash.sh (only
once, typically, but it should be idempotent).
Then run bash 10upgrade.sh and 20upgradeboth.sh and freely improve them
or change the code to see if this improves.

upgrade means "only libreant", while "upgradeboth" means "both libreant
and elasticsearch"

the most important bits are inside do-upgradeboth.sh for this case, yes.

I am sorry for the crazy naming, but except for the naming, the general
organization of code/tests shouldn't be that bad. So, do you had a hard
time understanding the code inside do-upgradeboth.sh ? I could try to
explain it more, but it seems pretty straightforward to me.

ael-code commented:
On 05/16/2016 09:00 PM, BoySka wrote:
> ael:
>>> Sure; added the debug flag, but there's not much more.
>>>
>>> - Logs
>>> - test "suite"
>>
>>I've tried hard, but I cannot understand which is the test case flow.
>>Did it happen in `do-upgradeboth.sh`? Can you provide a detailed test
>>flow?
> 
> sure. The numbered ones are thing I run on my system. The do-* ones are
> run inside the container.
> 
> 00base.sh, using do-base.sh, creates a base system for next
> tests: installs the stable version, ES 1.7, and creates some documents.
> the other tests are run on ephemeral containers, so they don't touch
> each other. The typical sequence for testing is bash 00bash.sh (only
> once, typically, but it should be idempotent).
> Then run bash 10upgrade.sh and 20upgradeboth.sh and freely improve them
> or change the code to see if this improves.
> 
> upgrade means "only libreant", while "upgradeboth" means "both libreant
> and elasticsearch"
> 
> the most important bits are inside do-upgradeboth.sh for this case, yes.
> 
> I am sorry for the crazy naming, but except for the naming, the general
> organization of code/tests shouldn't be that bad. So, do you had a hard
> time understanding the code inside do-upgradeboth.sh ? I could try to
> explain it more, but it seems pretty straightforward to me.
> 

1. You are completely crazy.
2. I hate bash for doing this stuff.
3. Sleep call are evil.


Anyway I've made some little adjustment to the test suite you are using.
You can find the code here [1].

Since you've asked, I've adjusted the `es_wait` function.

The upgrade-both test was failing also on my machine. But after I've
increased the sleep time after the libreant daemon start command
everything seems to work.

Can you check if this also the problem that you've raised till now?



[1] : https://gist.github.com/ael-code/993c24ecaec8fee1117162b34c2caf3a

boyska commented:
ael:
>On 05/16/2016 09:00 PM, BoySka wrote:
>> ael:
>>>> Sure; added the debug flag, but there's not much more.
>>>>
>>>> - Logs
>>>> - test "suite"
>>>
>>>I've tried hard, but I cannot understand which is the test case flow.
>>>Did it happen in `do-upgradeboth.sh`? Can you provide a detailed test
>>>flow?
>>
>> sure. The numbered ones are thing I run on my system. The do-* ones are
>> run inside the container.
>>
>> 00base.sh, using do-base.sh, creates a base system for next
>> tests: installs the stable version, ES 1.7, and creates some documents.
>> the other tests are run on ephemeral containers, so they don't touch
>> each other. The typical sequence for testing is bash 00bash.sh (only
>> once, typically, but it should be idempotent).
>> Then run bash 10upgrade.sh and 20upgradeboth.sh and freely improve them
>> or change the code to see if this improves.
>>
>> upgrade means "only libreant", while "upgradeboth" means "both libreant
>> and elasticsearch"
>>
>> the most important bits are inside do-upgradeboth.sh for this case, yes.
>>
>> I am sorry for the crazy naming, but except for the naming, the general
>> organization of code/tests shouldn't be that bad. So, do you had a hard
>> time understanding the code inside do-upgradeboth.sh ? I could try to
>> explain it more, but it seems pretty straightforward to me.
>>
>
>1. You are completely crazy.
>2. I hate bash for doing this stuff.

I agree with both :) but would you really have used something else?
I think it was good enough for this

>3. Sleep call are evil.

yep, that's why i asked about es_ready. Now that you fixed it, do you
think that the other sleep calls are very problematic in practice?

>The upgrade-both test was failing also on my machine. But after I've
>increased the sleep time after the libreant daemon start command
>everything seems to work.

btw, that's the kind of problems that should be solved with a
daemonizing process.
But that's ok for now.

For hard to explain network issues I have at the moment, I cannot. But
I'll try to do that asap.

ael-code commented:
> For hard to explain network issues I have at the moment, I cannot. But I'll try to do that asap.

Any news on this? can we proceed with the merge?

boyska commented:
I forgot what was it that I could not do :) however my memory and our comments confirm that this works, except that these timing issue make the test somewhat unreliable. But this is about the test itself, not about libreant code.
